File: src/sql/lexer.zig
//! SQL Lexer (Tokenizer)
//!
//! Converts SQL text into a stream of tokens for parsing.
//! Supports SELECT statements with WHERE, ORDER BY, LIMIT, GROUP BY, and vector search extensions.

const std = @import("std");

/// Token types for SQL lexer
pub const TokenType = enum {
    // Keywords
    SELECT,
    FROM,
    WHERE,
    AND,
    OR,
    NOT,
    IN,
    BETWEEN,
    LIKE,
    IS,
    NULL,
    AS,
    DISTINCT,
    ORDER,
    BY,
    ASC,
    DESC,
    LIMIT,
    OFFSET,
    GROUP,
    HAVING,

    // Vector search extension
    NEAR,
    TOPK,
    FILE,

    // Aggregate functions
    COUNT,
    SUM,
    AVG,
    MIN,
    MAX,

    // Literals
    IDENTIFIER,
    NUMBER,
    STRING,

    // Operators
    STAR,      // *
    EQ,        // =
    NE,        // !=, <>
    LT,        // <
    LE,        // <=
    GT,        // >
    GE,        // >=
    PLUS,      // +
    MINUS,     // -
    SLASH,     // /
    COMMA,     // ,
    DOT,       // .
    LPAREN,    // (
    RPAREN,    // )
    SEMICOLON, // ;

    EOF,
};

/// A token with its type and lexeme
pub const Token = struct {
    type: TokenType,
    lexeme: []const u8,
    position: usize,
};

/// SQL Lexer
pub const Lexer = struct {
    input: []const u8,
    position: usize,

    const Self = @This();

    /// Create a new lexer
    pub fn init(input: []const u8) Self {
        return Self{
            .input = input,
            .position = 0,
        };
    }

    /// Peek at current character without advancing
    fn peek(self: *const Self) ?u8 {
        if (self.position >= self.input.len) return null;
        return self.input[self.position];
    }

    /// Peek ahead n characters
    fn peekAhead(self: *const Self, n: usize) ?u8 {
        const pos = self.position + n;
        if (pos >= self.input.len) return null;
        return self.input[pos];
    }

    /// Advance and return current character
    fn advance(self: *Self) ?u8 {
        if (self.position >= self.input.len) return null;
        const ch = self.input[self.position];
        self.position += 1;
        return ch;
    }

    /// Skip whitespace
    fn skipWhitespace(self: *Self) void {
        while (self.peek()) |ch| {
            if (!std.ascii.isWhitespace(ch)) break;
            _ = self.advance();
        }
    }

    /// Read identifier or keyword
    fn readIdentifier(self: *Self) !Token {
        const start = self.position;

        while (self.peek()) |ch| {
            if (std.ascii.isAlphanumeric(ch) or ch == '_') {
                _ = self.advance();
            } else {
                break;
            }
        }

        const lexeme = self.input[start..self.position];
        const token_type = keywordOrIdentifier(lexeme);

        return Token{
            .type = token_type,
            .lexeme = lexeme,
            .position = start,
        };
    }

    /// Read numeric literal
    fn readNumber(self: *Self) !Token {
        const start = self.position;

        // Read digits
        while (self.peek()) |ch| {
            if (std.ascii.isDigit(ch)) {
                _ = self.advance();
            } else if (ch == '.') {
                // Handle decimal point
                _ = self.advance();
                // Read fractional part
                while (self.peek()) |digit| {
                    if (std.ascii.isDigit(digit)) {
                        _ = self.advance();
                    } else {
                        break;
                    }
                }
                break;
            } else {
                break;
            }
        }

        const lexeme = self.input[start..self.position];
        return Token{
            .type = .NUMBER,
            .lexeme = lexeme,
            .position = start,
        };
    }

    /// Read string literal (single or double quoted)
    fn readString(self: *Self, quote: u8) !Token {
        const start = self.position - 1; // Include opening quote

        while (self.peek()) |ch| {
            if (ch == quote) {
                _ = self.advance();
                break;
            } else if (ch == '\\') {
                // Handle escape sequences
                _ = self.advance();
                _ = self.advance(); // Skip escaped character
            } else {
                _ = self.advance();
            }
        }

        const lexeme = self.input[start..self.position];
        return Token{
            .type = .STRING,
            .lexeme = lexeme,
            .position = start,
        };
    }

    /// Get next token
    pub fn nextToken(self: *Self) !Token {
        self.skipWhitespace();

        const ch = self.peek() orelse {
            return Token{
                .type = .EOF,
                .lexeme = "",
                .position = self.position,
            };
        };

        const start = self.position;

        // Identifiers and keywords
        if (std.ascii.isAlphabetic(ch) or ch == '_') {
            return self.readIdentifier();
        }

        // Numbers
        if (std.ascii.isDigit(ch)) {
            return self.readNumber();
        }

        // Strings
        if (ch == '\'' or ch == '"') {
            _ = self.advance();
            return self.readString(ch);
        }

        // Operators and punctuation
        _ = self.advance();

        switch (ch) {
            '*' => return Token{ .type = .STAR, .lexeme = "*", .position = start },
            ',' => return Token{ .type = .COMMA, .lexeme = ",", .position = start },
            '.' => return Token{ .type = .DOT, .lexeme = ".", .position = start },
            '(' => return Token{ .type = .LPAREN, .lexeme = "(", .position = start },
            ')' => return Token{ .type = .RPAREN, .lexeme = ")", .position = start },
            ';' => return Token{ .type = .SEMICOLON, .lexeme = ";", .position = start },
            '+' => return Token{ .type = .PLUS, .lexeme = "+", .position = start },
            '-' => return Token{ .type = .MINUS, .lexeme = "-", .position = start },
            '/' => return Token{ .type = .SLASH, .lexeme = "/", .position = start },

            '=' => return Token{ .type = .EQ, .lexeme = "=", .position = start },

            '!' => {
                if (self.peek() == '=') {
                    _ = self.advance();
                    return Token{ .type = .NE, .lexeme = "!=", .position = start };
                }
                return error.UnexpectedCharacter;
            },

            '<' => {
                if (self.peek() == '=') {
                    _ = self.advance();
                    return Token{ .type = .LE, .lexeme = "<=", .position = start };
                } else if (self.peek() == '>') {
                    _ = self.advance();
                    return Token{ .type = .NE, .lexeme = "<>", .position = start };
                }
                return Token{ .type = .LT, .lexeme = "<", .position = start };
            },

            '>' => {
                if (self.peek() == '=') {
                    _ = self.advance();
                    return Token{ .type = .GE, .lexeme = ">=", .position = start };
                }
                return Token{ .type = .GT, .lexeme = ">", .position = start };
            },

            else => return error.UnexpectedCharacter,
        }
    }

    /// Tokenize entire input into array
    pub fn tokenize(self: *Self, allocator: std.mem.Allocator) ![]Token {
        var tokens = std.ArrayList(Token){};
        try tokens.ensureTotalCapacity(allocator, 32);
        errdefer tokens.deinit(allocator);

        while (true) {
            const token = try self.nextToken();
            try tokens.append(allocator, token);
            if (token.type == .EOF) break;
        }

        return tokens.toOwnedSlice(allocator);
    }
};

/// Map string to keyword or return IDENTIFIER
fn keywordOrIdentifier(word: []const u8) TokenType {
    // Convert to uppercase for case-insensitive comparison
    var upper_buf: [64]u8 = undefined;
    if (word.len > upper_buf.len) return .IDENTIFIER;

    for (word, 0..) |ch, i| {
        upper_buf[i] = std.ascii.toUpper(ch);
    }
    const upper = upper_buf[0..word.len];

    // Keywords (alphabetically sorted for binary search)
    if (std.mem.eql(u8, upper, "AND")) return .AND;
    if (std.mem.eql(u8, upper, "AS")) return .AS;
    if (std.mem.eql(u8, upper, "ASC")) return .ASC;
    if (std.mem.eql(u8, upper, "AVG")) return .AVG;
    if (std.mem.eql(u8, upper, "BETWEEN")) return .BETWEEN;
    if (std.mem.eql(u8, upper, "BY")) return .BY;
    if (std.mem.eql(u8, upper, "COUNT")) return .COUNT;
    if (std.mem.eql(u8, upper, "DESC")) return .DESC;
    if (std.mem.eql(u8, upper, "DISTINCT")) return .DISTINCT;
    if (std.mem.eql(u8, upper, "FILE")) return .FILE;
    if (std.mem.eql(u8, upper, "FROM")) return .FROM;
    if (std.mem.eql(u8, upper, "GROUP")) return .GROUP;
    if (std.mem.eql(u8, upper, "HAVING")) return .HAVING;
    if (std.mem.eql(u8, upper, "IN")) return .IN;
    if (std.mem.eql(u8, upper, "IS")) return .IS;
    if (std.mem.eql(u8, upper, "LIKE")) return .LIKE;
    if (std.mem.eql(u8, upper, "LIMIT")) return .LIMIT;
    if (std.mem.eql(u8, upper, "MAX")) return .MAX;
    if (std.mem.eql(u8, upper, "MIN")) return .MIN;
    if (std.mem.eql(u8, upper, "NEAR")) return .NEAR;
    if (std.mem.eql(u8, upper, "NOT")) return .NOT;
    if (std.mem.eql(u8, upper, "NULL")) return .NULL;
    if (std.mem.eql(u8, upper, "OFFSET")) return .OFFSET;
    if (std.mem.eql(u8, upper, "OR")) return .OR;
    if (std.mem.eql(u8, upper, "ORDER")) return .ORDER;
    if (std.mem.eql(u8, upper, "SELECT")) return .SELECT;
    if (std.mem.eql(u8, upper, "SUM")) return .SUM;
    if (std.mem.eql(u8, upper, "TOPK")) return .TOPK;
    if (std.mem.eql(u8, upper, "WHERE")) return .WHERE;

    return .IDENTIFIER;
}

// ============================================================================
// Tests
// ============================================================================

test "lexer basic tokens" {
    const sql = "SELECT * FROM table WHERE id = 42";
    var lexer = Lexer.init(sql);

    const allocator = std.testing.allocator;
    const tokens = try lexer.tokenize(allocator);
    defer allocator.free(tokens);

    try std.testing.expectEqual(TokenType.SELECT, tokens[0].type);
    try std.testing.expectEqual(TokenType.STAR, tokens[1].type);
    try std.testing.expectEqual(TokenType.FROM, tokens[2].type);
    try std.testing.expectEqual(TokenType.IDENTIFIER, tokens[3].type);
    try std.testing.expectEqual(TokenType.WHERE, tokens[4].type);
    try std.testing.expectEqual(TokenType.IDENTIFIER, tokens[5].type);
    try std.testing.expectEqual(TokenType.EQ, tokens[6].type);
    try std.testing.expectEqual(TokenType.NUMBER, tokens[7].type);
}

test "lexer string literals" {
    const sql = "SELECT 'hello' FROM \"world\"";
    var lexer = Lexer.init(sql);

    const allocator = std.testing.allocator;
    const tokens = try lexer.tokenize(allocator);
    defer allocator.free(tokens);

    try std.testing.expectEqual(TokenType.STRING, tokens[1].type);
    try std.testing.expectEqual(TokenType.STRING, tokens[3].type);
}

test "lexer comparison operators" {
    const sql = "a != b AND c <> d AND e <= f AND g >= h";
    var lexer = Lexer.init(sql);

    const allocator = std.testing.allocator;
    const tokens = try lexer.tokenize(allocator);
    defer allocator.free(tokens);

    try std.testing.expectEqual(TokenType.NE, tokens[1].type);
    try std.testing.expectEqual(TokenType.NE, tokens[5].type);
    try std.testing.expectEqual(TokenType.LE, tokens[9].type);
    try std.testing.expectEqual(TokenType.GE, tokens[13].type);
}

File: packages/browser/src/client/sql/lexer.js
/**
 * SQLLexer - SQL tokenization
 */

const TokenType = {
    // Keywords
    SELECT: 'SELECT', DISTINCT: 'DISTINCT', FROM: 'FROM', WHERE: 'WHERE',
    AND: 'AND', OR: 'OR', NOT: 'NOT', ORDER: 'ORDER', BY: 'BY',
    ASC: 'ASC', DESC: 'DESC', LIMIT: 'LIMIT', OFFSET: 'OFFSET', AS: 'AS',
    NULL: 'NULL', IS: 'IS', IN: 'IN', BETWEEN: 'BETWEEN', LIKE: 'LIKE',
    TRUE: 'TRUE', FALSE: 'FALSE', GROUP: 'GROUP', HAVING: 'HAVING',
    QUALIFY: 'QUALIFY', ROLLUP: 'ROLLUP', CUBE: 'CUBE', GROUPING: 'GROUPING', SETS: 'SETS',
    COUNT: 'COUNT', SUM: 'SUM', AVG: 'AVG', MIN: 'MIN', MAX: 'MAX',
    NEAR: 'NEAR', TOPK: 'TOPK', FILE: 'FILE',
    JOIN: 'JOIN', INNER: 'INNER', LEFT: 'LEFT', RIGHT: 'RIGHT', FULL: 'FULL',
    OUTER: 'OUTER', CROSS: 'CROSS', ON: 'ON',
    CREATE: 'CREATE', TABLE: 'TABLE', INSERT: 'INSERT', INTO: 'INTO', VALUES: 'VALUES',
    UPDATE: 'UPDATE', SET: 'SET', DELETE: 'DELETE', DROP: 'DROP', IF: 'IF', EXISTS: 'EXISTS',
    INT: 'INT', INTEGER: 'INTEGER', BIGINT: 'BIGINT', FLOAT: 'FLOAT', REAL: 'REAL',
    DOUBLE: 'DOUBLE', TEXT: 'TEXT', VARCHAR: 'VARCHAR', BOOLEAN: 'BOOLEAN', BOOL: 'BOOL',
    VECTOR: 'VECTOR', PRIMARY: 'PRIMARY', KEY: 'KEY',
    WITH: 'WITH', RECURSIVE: 'RECURSIVE', UNION: 'UNION', ALL: 'ALL',
    PIVOT: 'PIVOT', UNPIVOT: 'UNPIVOT', FOR: 'FOR',
    INTERSECT: 'INTERSECT', EXCEPT: 'EXCEPT',
    OVER: 'OVER', PARTITION: 'PARTITION', ROW_NUMBER: 'ROW_NUMBER', RANK: 'RANK',
    DENSE_RANK: 'DENSE_RANK', NTILE: 'NTILE', LAG: 'LAG', LEAD: 'LEAD',
    FIRST_VALUE: 'FIRST_VALUE', LAST_VALUE: 'LAST_VALUE', NTH_VALUE: 'NTH_VALUE',
    PERCENT_RANK: 'PERCENT_RANK', CUME_DIST: 'CUME_DIST',
    ROWS: 'ROWS', RANGE: 'RANGE', UNBOUNDED: 'UNBOUNDED', PRECEDING: 'PRECEDING',
    FOLLOWING: 'FOLLOWING', CURRENT: 'CURRENT', ROW: 'ROW',
    EXPLAIN: 'EXPLAIN', ARRAY: 'ARRAY', CASE: 'CASE', WHEN: 'WHEN', THEN: 'THEN',
    ELSE: 'ELSE', END: 'END', CAST: 'CAST', COALESCE: 'COALESCE', NULLIF: 'NULLIF',
    // Literals & Operators
    IDENTIFIER: 'IDENTIFIER', NUMBER: 'NUMBER', STRING: 'STRING',
    STAR: 'STAR', COMMA: 'COMMA', DOT: 'DOT', LPAREN: 'LPAREN', RPAREN: 'RPAREN',
    EQ: 'EQ', NE: 'NE', LT: 'LT', LE: 'LE', GT: 'GT', GE: 'GE',
    PLUS: 'PLUS', MINUS: 'MINUS', SLASH: 'SLASH', LBRACKET: 'LBRACKET', RBRACKET: 'RBRACKET',
    EOF: 'EOF',
};

const KEYWORDS = {
    'SELECT': TokenType.SELECT, 'DISTINCT': TokenType.DISTINCT, 'FROM': TokenType.FROM,
    'WHERE': TokenType.WHERE, 'AND': TokenType.AND, 'OR': TokenType.OR, 'NOT': TokenType.NOT,
    'ORDER': TokenType.ORDER, 'BY': TokenType.BY, 'ASC': TokenType.ASC, 'DESC': TokenType.DESC,
    'LIMIT': TokenType.LIMIT, 'OFFSET': TokenType.OFFSET, 'AS': TokenType.AS,
    'NULL': TokenType.NULL, 'IS': TokenType.IS, 'IN': TokenType.IN, 'BETWEEN': TokenType.BETWEEN,
    'LIKE': TokenType.LIKE, 'TRUE': TokenType.TRUE, 'FALSE': TokenType.FALSE,
    'GROUP': TokenType.GROUP, 'HAVING': TokenType.HAVING, 'QUALIFY': TokenType.QUALIFY,
    'ROLLUP': TokenType.ROLLUP, 'CUBE': TokenType.CUBE, 'GROUPING': TokenType.GROUPING, 'SETS': TokenType.SETS,
    'COUNT': TokenType.COUNT, 'SUM': TokenType.SUM, 'AVG': TokenType.AVG, 'MIN': TokenType.MIN, 'MAX': TokenType.MAX,
    'NEAR': TokenType.NEAR, 'TOPK': TokenType.TOPK, 'FILE': TokenType.FILE,
    'JOIN': TokenType.JOIN, 'INNER': TokenType.INNER, 'LEFT': TokenType.LEFT, 'RIGHT': TokenType.RIGHT,
    'FULL': TokenType.FULL, 'OUTER': TokenType.OUTER, 'CROSS': TokenType.CROSS, 'ON': TokenType.ON,
    'CREATE': TokenType.CREATE, 'TABLE': TokenType.TABLE, 'INSERT': TokenType.INSERT, 'INTO': TokenType.INTO,
    'VALUES': TokenType.VALUES, 'UPDATE': TokenType.UPDATE, 'SET': TokenType.SET, 'DELETE': TokenType.DELETE,
    'DROP': TokenType.DROP, 'IF': TokenType.IF, 'EXISTS': TokenType.EXISTS,
    'INT': TokenType.INT, 'INTEGER': TokenType.INTEGER, 'BIGINT': TokenType.BIGINT,
    'FLOAT': TokenType.FLOAT, 'REAL': TokenType.REAL, 'DOUBLE': TokenType.DOUBLE,
    'TEXT': TokenType.TEXT, 'VARCHAR': TokenType.VARCHAR, 'BOOLEAN': TokenType.BOOLEAN, 'BOOL': TokenType.BOOL,
    'VECTOR': TokenType.VECTOR, 'PRIMARY': TokenType.PRIMARY, 'KEY': TokenType.KEY,
    'WITH': TokenType.WITH, 'RECURSIVE': TokenType.RECURSIVE, 'UNION': TokenType.UNION, 'ALL': TokenType.ALL,
    'PIVOT': TokenType.PIVOT, 'UNPIVOT': TokenType.UNPIVOT, 'FOR': TokenType.FOR,
    'INTERSECT': TokenType.INTERSECT, 'EXCEPT': TokenType.EXCEPT,
    'OVER': TokenType.OVER, 'PARTITION': TokenType.PARTITION, 'ROW_NUMBER': TokenType.ROW_NUMBER,
    'RANK': TokenType.RANK, 'DENSE_RANK': TokenType.DENSE_RANK, 'NTILE': TokenType.NTILE,
    'LAG': TokenType.LAG, 'LEAD': TokenType.LEAD, 'FIRST_VALUE': TokenType.FIRST_VALUE,
    'LAST_VALUE': TokenType.LAST_VALUE, 'NTH_VALUE': TokenType.NTH_VALUE,
    'PERCENT_RANK': TokenType.PERCENT_RANK, 'CUME_DIST': TokenType.CUME_DIST,
    'ROWS': TokenType.ROWS, 'RANGE': TokenType.RANGE, 'UNBOUNDED': TokenType.UNBOUNDED,
    'PRECEDING': TokenType.PRECEDING, 'FOLLOWING': TokenType.FOLLOWING,
    'CURRENT': TokenType.CURRENT, 'ROW': TokenType.ROW,
    'EXPLAIN': TokenType.EXPLAIN, 'ARRAY': TokenType.ARRAY,
    'CASE': TokenType.CASE, 'WHEN': TokenType.WHEN, 'THEN': TokenType.THEN,
    'ELSE': TokenType.ELSE, 'END': TokenType.END, 'CAST': TokenType.CAST,
    'COALESCE': TokenType.COALESCE, 'NULLIF': TokenType.NULLIF,
};

class SQLLexer {
    constructor(sql) {
        this.sql = sql;
        this.pos = 0;
        this.length = sql.length;
    }

    peek() {
        if (this.pos >= this.length) return '\0';
        return this.sql[this.pos];
    }

    advance() {
        if (this.pos < this.length) {
            return this.sql[this.pos++];
        }
        return '\0';
    }

    skipWhitespace() {
        while (this.pos < this.length && /\s/.test(this.sql[this.pos])) {
            this.pos++;
        }
    }

    readIdentifier() {
        const start = this.pos;
        while (this.pos < this.length && /[a-zA-Z0-9_]/.test(this.sql[this.pos])) {
            this.pos++;
        }
        return this.sql.slice(start, this.pos);
    }

    readNumber() {
        const start = this.pos;
        let hasDecimal = false;

        while (this.pos < this.length) {
            const ch = this.sql[this.pos];
            if (ch === '.' && !hasDecimal) {
                hasDecimal = true;
                this.pos++;
            } else if (/\d/.test(ch)) {
                this.pos++;
            } else {
                break;
            }
        }
        return this.sql.slice(start, this.pos);
    }

    readString(quote) {
        const start = this.pos;
        this.advance(); // Skip opening quote

        while (this.pos < this.length) {
            const ch = this.sql[this.pos];
            if (ch === quote) {
                // Check for escaped quote
                if (this.pos + 1 < this.length && this.sql[this.pos + 1] === quote) {
                    this.pos += 2;
                    continue;
                }
                this.pos++; // Skip closing quote
                break;
            }
            this.pos++;
        }

        // Return string without quotes, handling escaped quotes
        const inner = this.sql.slice(start + 1, this.pos - 1);
        return inner.replace(new RegExp(quote + quote, 'g'), quote);
    }

    nextToken() {
        this.skipWhitespace();

        if (this.pos >= this.length) {
            return { type: TokenType.EOF, value: null };
        }

        const ch = this.peek();

        // Identifiers and keywords
        if (/[a-zA-Z_]/.test(ch)) {
            const value = this.readIdentifier();
            const upper = value.toUpperCase();
            const type = KEYWORDS[upper] || TokenType.IDENTIFIER;
            return { type, value: type === TokenType.IDENTIFIER ? value : upper };
        }

        // Numbers
        if (/\d/.test(ch)) {
            const value = this.readNumber();
            return { type: TokenType.NUMBER, value };
        }

        // Strings
        if (ch === "'" || ch === '"') {
            const value = this.readString(ch);
            return { type: TokenType.STRING, value };
        }

        // Operators
        this.advance();

        switch (ch) {
            case '*': return { type: TokenType.STAR, value: '*' };
            case ',': return { type: TokenType.COMMA, value: ',' };
            case '.': return { type: TokenType.DOT, value: '.' };
            case '(': return { type: TokenType.LPAREN, value: '(' };
            case ')': return { type: TokenType.RPAREN, value: ')' };
            case '+': return { type: TokenType.PLUS, value: '+' };
            case '-': return { type: TokenType.MINUS, value: '-' };
            case '/': return { type: TokenType.SLASH, value: '/' };
            case '[': return { type: TokenType.LBRACKET, value: '[' };
            case ']': return { type: TokenType.RBRACKET, value: ']' };
            case '=': return { type: TokenType.EQ, value: '=' };
            case '<':
                if (this.peek() === '=') {
                    this.advance();
                    return { type: TokenType.LE, value: '<=' };
                }
                if (this.peek() === '>') {
                    this.advance();
                    return { type: TokenType.NE, value: '<>' };
                }
                return { type: TokenType.LT, value: '<' };
            case '>':
                if (this.peek() === '=') {
                    this.advance();
                    return { type: TokenType.GE, value: '>=' };
                }
                return { type: TokenType.GT, value: '>' };
            case '!':
                if (this.peek() === '=') {
                    this.advance();
                    return { type: TokenType.NE, value: '!=' };
                }
                throw new Error(`Unexpected character: ${ch}`);
            default:
                throw new Error(`Unexpected character: ${ch}`);
        }
    }

    tokenize() {
        const tokens = [];
        let token;
        while ((token = this.nextToken()).type !== TokenType.EOF) {
            tokens.push(token);
        }
        tokens.push(token); // Include EOF
        return tokens;
    }
}

/**
 * SQL Parser - parses tokens into AST
 */

export { SQLLexer, TokenType, KEYWORDS };
